src/aspire/source/__init__.py

```diff
@@ -236,52 +324,56 @@ def cache(self, im=None):
             im = self.images()
         self._im = im
 
-    def images(self, start=0, num=np.inf, indices=None, *args, **kwargs):
+    def images(self, start=0, num=np.inf, indices=None, batch_size=256, *args, **kwargs):
         if indices is None:
             indices = np.arange(start, min(start + num, self.n))
 
         if self._im is not None:
+            logger.info(f'Loading images from cache')
             im = Image(self._im[:, :, indices])
         else:
-            im = self._images(start=start, num=num, indices=indices, *args, **kwargs)
-
+            resolution = min(self.generation_pipeline.resolution, self.L)
+            im = np.empty((resolution, resolution, len(indices)), dtype=self.dtype)
+            for i in range(0, len(indices), batch_size):
+                chunk_range = np.arange(i, min(i + batch_size, len(indices)))
+                chunk_indices = indices[chunk_range]
+                _im = self._images(start=np.min(chunk_indices), indices=chunk_indices, *args, **kwargs)
+                _im = self.generation_pipeline.forward(_im, indices=chunk_indices)
+                im[:, :, chunk_range] = _im.asnumpy()
+            im = Image(im)
+
+        logger.info(f'Loaded {len(indices)} images')
         return im
 
-    def set_max_resolution(self, max_L):
+    def downsample(self, max_L):
         ensure(max_L <= self.L, "Max desired resolution should be less than the current resolution")
         logger.info(f'Setting max. resolution of source = {max_L}')
-        self.L = max_L
 
-        ds_factor = self._L / max_L
-        self.filters = [f.scale(ds_factor) for f in self.filters]
-        self.offsets /= ds_factor
+        self.model_pipeline.downsample(resolution=max_L)
+        self.generation_pipeline.add_transform(DownSample(resolution=max_L))
 
+        self.L = max_L
         # Invalidate images
         self._im = None
 
-    def whiten(self, whiten_filter=None):
+    def whiten(self, whiten_filter):
         """
         Modify the Source object in place by whitening + caching all images, and adding the appropriate whitening
             filter to all available filters.
-        :param whiten_filter: Whitening filter to apply. If None, determined automatically.
+        :param whiten_filter: Whitening filter to apply.
```

Comment:
Why not allow `ImageSource` to estimate its own whitening filter?

Reply:
I thought it was a little random for the whitening operation to assume white noise, but maybe not. Added back in.

Reply:
Well it's not really appropriate for it to assume white noise (where the noise PSD is constant, and it would not need any whitening), but something like a radially symmetric noise PSD might be reasonable. If that's not available yet, let's leave it without a default for now.

Reply:
Default value of `whiten_filter` removed.